### PR TITLE
Created Dockerfile and updated ReadMe

### LINKED ISSUE #120: Local Docker setup
Would be nice to have a `dockerfile` and local Docker dev instructions in the README

### PR DESCRIPTION
**Pull request by Veinar**
Hi,

Briefly, I think i could resolve #120.

I've created basic `Dockerfile` that uses `alpine` image as base - just to keep it as lightweight as possible.
Then I added an environment variable without which the build would not pass `NODE_OPTIONS=--openssl-legacy-provider`.
In the next step, I install the necessary dependencies using the `apk` command, in addition, I made sure to have a git on board in case I need to download something auxiliary.
Before copying application files I create workdir (`gitops-website`) to store it in fixed place.
After that Docker will copy all files from local filesystem. I didn't create a `.dockerignore` file because I'm not an expert on node applications and I don't know what might not be added to the final image. (I assumed everything was needed).
Lastly I expose port `8000` on which node by default exposes running application, and created entry point accordingly to ReadMe file.

**I've had some issues with exposing application, and then I discovered that command:**
`gatsby develop` needs additional parameter (`--host=0.0.0.0`) in order to listen on all interfaces which was required to reach application from outside of docker container. So I've made changes in `package.json` and added this parameter into line with `gatsby develop` as you can see in commit. Now it is only required to pass `0.0.0.0` during port mapping in `docker run`.

Of course, I added information in the ReadMe file on how to build the image and how to run it.

Hope I helped,
Waiting for your review,
Regards,
Konrad 🙂

### CODE REVIEW

File: package.json
Comment (christianh814) [inline package.json:12]: How does this æffect running it in production? (if at all) @niklasmtj

File: Dockerfile
Comment (christianh814) [inline Dockerfile:20]: Talked to @samueltauil about this. I would rather not run it as root if at all possible.

I would suggest adding something like

```
RUN chown -R 1001:0 /gitops-website && chmod -R ug+rwx /gitops-website
USER 1001
```

**Review by christianh814 — CHANGES_REQUESTED** [review]
In general I like this. But I have a few questions concerns before I can approve.

- Does this emulate the env that the site is running on? (I'm mainly concerned about the version of node)
- The Container would run as root. Which isn't terrible if you're just developing locally but we should probably do the best practice here.
- Looks like we need to change the `package.json` file. What is the implications?
- DCO isn't passing.

File: Dockerfile
Comment (niklasmtj) [reply Dockerfile:20]: Maybe we can even use `node:16-alpine`?

File: package.json
Comment (niklasmtj) [reply package.json:12]: This shouldn't effect it at all since we're pushing the public folder to GH pages

**Comment by samueltauil** [comment]
I would suggest changing the name of the file from Dockerfile to Containerfile as the file where you define the instructions for building an OCI-compliant container image is traditionally named Dockerfile due to its widespread use and association with Docker, however, to avoid confusion or to align with a different ecosystem, some tools or environments might opt to use a different naming convention like Containerfile, especially if they're aiming to support multiple container runtimes or be more agnostic towards the containerization framework. 

I would also think of the Containerfile definition as this: 

```
FROM node:latest

# Create a non-root user
RUN addgroup -g 1001 appuser && adduser -S -u 1001 -G appuser appuser

# Setup work directory and permissions for non-root user
WORKDIR /gitops-website
RUN chown -R appuser:appuser /gitops-website

# Switch to non-root user
USER appuser

# Add required files
COPY --chown=appuser:appuser . .

# Install dependencies
RUN npm ci --only=production

# Expose port for server
EXPOSE 8000

# Create entrypoint
CMD ["npm", "start"]
```

**Comment by Veinar** [comment]
Okay I will refer to @christianh814 questions, and take suggestions from @samueltauil in order to prepare updated version. I was using alpine image as a base which can be problematic when there is a need to specify node version.

**Comment by Veinar** [comment]
Hi,

I've updated PR with latest commit 🙂 

Now as I commented earlier will refer to questions from @christianh814,

* Does this emulate the env that the site is running on? (I'm mainly concerned about the version of node)
> I do not have information about current setup, but regarding node version I've made changes within `Containerfile` where you now specify `NODE_VERSION` in build which selects version from `node` docker image. 
* The Container would run as root. Which isn't terrible if you're just developing locally but we should probably do the best practice here.
> Yes totally agree, changed to built-in user `node`. Slightly different than @samueltauil proposed in his draft but It might be enough.
* Looks like we need to change the package.json file. What is the implications?
> To the best of my knowledge, there will be no implications, because adding `--host=0.0.0.0` only sets the listener to a wider range. A similar solution is used, for example, if you want to reach an application from a local network from a second device. This should be considered more as a door ajar a bit more rather than a change that will affect the operation of the application.
* DCO isn't passing.
> Yes I am aware of that, and it is caused because my first commit was not signed, in second one I've corrected myself but used SSH key, which does not meet your standards (for which I am very sorry 🙏 ), my default signature is using SSH rather than `Signed-off-by: Author Name <authoremail@example.com>`

Refering to comment from @samueltauil,

I've changed name as required 🙂  and used node image as base, but I was taught to use as lightweight image as possible so I'm always trying to use image that bases on minimal image possible.

What's more I've changed `docker run` port mapping as `0.0.0.0:80` is now not required when using `node` image as base.

If you have any other ideas/requests for me to change, please let me know,
Have nice weekend 👋 

Regards,
Konrad

**Comment by niklasmtj** [comment]
Hey @Veinar, thanks a lot for that PR! I'll have a look later today. 
One question: Did you make it run with Node19 yet? The GitHub Actions uses 16 but using 18 or 20 would be better in the near future.

**Comment by Veinar** [comment]
Hi @niklasmtj, thanks for response 🙏 

I'm not a professional tester, but I have successfully built containers using as a base image:

- `16-alpine` ✔️ 
- `18-alpine` ✔️ 
- `19-alpine` ✔️ 
- `20-alpine` ✔️ 
- `21-alpine` ✔️ 

And I was able to display website using port mapping (`-p 80:8000`) on my local PC 🙂
As said before I'm not a tester but for me it looked fine/same (comparing as it was run on `node:16-alpine`). **If you want I could make additional commit that will change default node version** in build that `--NODE_VERSION` will not be required argument during build. It would rather be additional customization. 

Please let me know,
Regards,
Konrad

P.S: Possible versions of base image could be withdrawn using this command (bash):
```bash
wget -q -O - "https://hub.docker.com/v2/namespaces/library/repositories/node/tags?page_size=1000" | jq -r '.results[] | select(.name | endswith("-alpine")) | .name' | sort -r
```
P.S2: To use by default latest version of `node` image, inc code `arg` value should be set to `current-alpine`. 🙂

File: Containerfile
Comment (niklasmtj) [inline Containerfile:2]:
```suggestion
ARG NODE_VERSION=20
```

I would suggest setting a default value here so we can also build it without the build argument if we want. What do you think?

File: Containerfile
Comment (niklasmtj) [inline Containerfile:8]: This line seems to be the reason I get errors when building it with node 16. Starting with node 18 this doesn't seem to bother the builder so I'm fine with that.

Error for reference:
```
 > [6/6] RUN npm ci:
#11 0.182 node: --openssl-legacy-provider is not allowed in NODE_OPTIONS
```

File: README.md
Comment (niklasmtj) [inline README.md:23]:
```suggestion
# Build container image from local filesystem with default Node version
docker build --no-cache -t website:<tag> -f Containerfile .

# Build container image from local filesystem with Node version as build arg
docker build --build-arg NODE_VERSION=19 --no-cache -t website:<tag> -f Containerfile .
```

This would make sense when we put a default value to in the `Containerfile`.

File: README.md
Comment (niklasmtj) [inline README.md:26]:
```suggestion
docker run -dit -p 8000:8000 website:<tag>
```
I understand why you default to 80 here and it also makes a lot of sense. What I think is that this might confuse folks since the console output for example says the following:
```
You can now view open-git-ops in the browser.
⠀
  Local:            http://localhost:8000/
```

File: Containerfile
Comment (niklasmtj) [inline Containerfile:23]: We need a `.dockerignore` file with `node_modules` in it. Otherwise we copy the local `node_modules` directory into the container which can result in flaky `npm ci` since we also then have possible development dependencies inside of the container.
@samueltauil is there also a "new" or more general name for `.dockerignore` like `.containerignore` that all the builder tools understand?

**Review by niklasmtj — COMMENTED** [review]
Hey Konrad (@Veinar),

This looks like a great addition. Thanks again for this!
I got a few comments and suggestions on your changes after testing it locally. Still, I'm pretty happy with these changes!

File: Containerfile
Comment (niklasmtj) [reply Containerfile:2]: Ah before i forget. I used node-20 since this is the latest LTS version

File: Containerfile
Comment (Veinar) [reply Containerfile:8]: Yes I forgot to mention that setting should be ommited in version 16, my bad.

File: Containerfile
Comment (Veinar) [reply Containerfile:23]: I'm wondering whether to use the `.dockerignore` name, which will be useful if a potential container user tries to build locally with docker/podman. And as @samueltauil said, the engine also checks this file name.

**Comment by Veinar** [comment]
Hi @niklasmtj,

I made the changes based on your suggestions, thank you for picking out the node version that should be the default base for the application. In addition, I included in the comment when adding the `ENV` that this env should not be set when using node version 16 or lower. ℹ️ 

As I wrote in the comment for `.containerignore` I would prefer to stay with the name `.dockerignore` due to the fact that when building an image locally via the `docker` or `podman` command - as far as I know - I will not be able to specify the name of the file with objects to ignore (for example, `.containerignore`) during calling the build command. Naming when using these types of files is explained here (for local `docker build`): 
https://docs.docker.com/build/building/context/#filename-and-location

Regards,
Konrad 🙂

File: Containerfile
Comment (niklasmtj) [reply Containerfile:8]: Looks good!

File: Containerfile
Comment (niklasmtj) [reply Containerfile:23]: I would be fine running with `.dockerignore` for now.

**Review by niklasmtj — APPROVED** [review]
LGTM. Let's see what the other maintainers say ☺️ 

Thank you very much Konrad!

**Comment by Veinar** [comment]
Hi team,

any update on this ? @christianh814 or @samueltauil.
I don't want to be rude, but this pull request is dangling.

Regards,
Konrad

**Review by christianh814 — APPROVED** [review]
LGTM!

**Comment by samueltauil** [comment]
LGTM! with a note that I still support the usage of `.containerignore` as all other tools support both options except for Docker.

**Comment by Veinar** [comment]
Thank you guys! 🙏 

Konrad

**Comment by niklasmtj** [comment]
> Thank you guys! 🙏 
> 
> 
> 
> Konrad

@Veinar thank you very much for your quick help! Really appreciate it

**Pull request merged**
